src/NHibernate/Util/DelegateHelper.cs

```diff
@@ -19,9 +19,15 @@ internal static class DelegateHelper
 			var parameter = Expression.Parameter(typeof(object), "x");
 			var instance = Expression.Convert(parameter, type);
 			var property = Expression.Property(instance, propertyName);
-			var value = Expression.Parameter(typeof (T), "value");
+			var valueParameter = Expression.Parameter(typeof (T), "value");
+			Expression value = valueParameter;
+			// Cast value if required
+			if (!property.Type.IsAssignableFrom(typeof(T)))
```

Comment:
Changes required for supporting setting without being able to provide the actual property type as a generic parameter.